frontend/js/src/common/brainzplayer/BrainzPlayer.tsx

```diff
@@ -487,82 +529,197 @@ export default class BrainzPlayer extends React.Component<
       !isListenFromDatasource(listen, datasource) &&
       !datasource.canSearchAndPlayTracks()
     ) {
-      this.playListen(listen, datasourceIndex + 1);
+      playListen(listen, nextListenIndex, datasourceIndex + 1);
       return;
     }
-    this.stopOtherBrainzPlayers();
-    this.setState({ currentDataSourceIndex: selectedDatasourceIndex }, () => {
+    stopOtherBrainzPlayers();
+    dispatch({ currentDataSourceIndex: selectedDatasourceIndex }, async () => {
+      while (
+        currentListenRef.current !== listen ||
+        currentDataSourceIndexRef.current !== selectedDatasourceIndex
+      ) {
+        // eslint-disable-next-line no-await-in-loop, no-promise-executor-return
+        await new Promise((resolve) => setTimeout(resolve, 100));
+      }
       datasource.playListen(listen);
     });
   };
 
-  togglePlay = async (): Promise<void> => {
-    try {
-      const { currentDataSourceIndex, playerPaused } = this.state;
-      const dataSource = this.dataSources[currentDataSourceIndex]?.current;
-      if (!dataSource) {
-        this.invalidateDataSource();
+  const playNextTrack = (invert: boolean = false): void => {
+    if (!isActivatedRef.current) {
+      // Player has not been activated by the user, do nothing.
+      return;
+    }
+    debouncedCheckProgressAndSubmitListen.flush();
+
+    const currentQueue = queueRef.current;
+    const currentAmbientQueue = ambientQueueRef.current;
+
+    if (currentQueue.length === 0 && currentAmbientQueue.length === 0) {
+      handleWarning(
+        "You can try loading listens or refreshing the page",
+        "No listens to play"
+      );
+      return;
+    }
+
+    const currentPlayingListenIndex = currentListenIndexRef.current;
+
+    let nextListenIndex: number;
+    if (queueRepeatMode === QueueRepeatModes.one) {
+      nextListenIndex =
+        currentPlayingListenIndex + (currentPlayingListenIndex < 0 ? 1 : 0);
+    } else {
+      nextListenIndex = currentPlayingListenIndex + (invert === true ? -1 : 1);
+    }
+
+    if (nextListenIndex < 0) {
+      nextListenIndex = currentQueue.length - 1;
+    }
+
+    if (nextListenIndex < currentQueue.length) {
+      const nextListen = currentQueue[nextListenIndex];
+      playListen(nextListen, nextListenIndex, 0);
+      return;
+    }
+
+    // If nextListenIndex exceeds the queue length, then pop the first item from the ambient queue, if available
+    // and add it to the end of the queue to play it next, otherwise wrap around to the first track
+    if (currentAmbientQueue.length > 0) {
+      const ambientQueueTop = currentAmbientQueue.shift();
+      if (ambientQueueTop) {
```

Comment:
Cool! I've added some comments to make it easier to understand how it's working right now. 

Yes, it means that we're at the end of queue. So we're checking if there are any tracks in the ambient queue, if yes, we'll play the next item, otherwise we'll wrap around and start playing the first track from the queue.